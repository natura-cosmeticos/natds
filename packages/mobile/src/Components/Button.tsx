import * as React from 'react';
import { Button as PaperButton, withTheme } from 'react-native-paper';
import { Surface, StyleProp, ViewStyle, TextStyle } from 'react-native';
import { IconSource } from 'react-native-paper/lib/typescript/src/components/Icon';
import { IThemeShape } from 'Provider/IThemeShape';
import { IButtonSize } from '@naturacosmeticos/natds-styles';

type ColorType = 'primary' | 'secondary' | 'default';
type ButtonSize = 'small' | 'medium' | 'large';

declare type ButtonProps = React.ComponentProps<typeof Surface> & {
  /**
   * Custom flag to change color.
   */
  colorType?: ColorType;
  /**
   * Mode of the button. You can change the mode to adjust the styling to give it desired emphasis.
   * - `text` - flat button without background or outline (low emphasis)
   * - `outlined` - button with an outline (medium emphasis)
   * - `contained` - button with a background color and elevation shadow (high emphasis)
   */
  mode?: 'text' | 'outlined' | 'contained';
  /**
   * Custom text color for flat button, or background color for contained button.
   */
  color?: string;
  /**
   * Icon to display for the `Button`.
   */
  icon?: IconSource;
  /**
   * Whether the button is disabled. A disabled button is greyed out and `onPress` is not called on touch.
   */
  disabled?: boolean;
  /**
   * Label text of the button.
   */
  children: React.ReactNode;
  /**
   * Make the label text uppercased. Note that this won't work if you pass React elements as children.
   */
  uppercase?: boolean;
  /**
   * Accessibility label for the button. This is read by the screen reader when the user taps the button.
   */
  accessibilityLabel?: string;
  /**
   * Function to execute on press.
   */
  onPress?: () => void;
  /**
   * Style of button's inner content.
   * Use this prop to apply custom height and width.
   */
  contentStyle?: StyleProp<ViewStyle>;
  /**
   * React Native style object.
   */
  style?: StyleProp<ViewStyle>;
  /**
   * Style for the button text.
   */
  labelStyle?: StyleProp<TextStyle>;
  /**
   * @optional
   */
  theme: IThemeShape;
  /**
   * Shadow depth, using to ios/web.
   */
  elevation?: number;
  /**
   * Default sizes (small, medium and large).
   */
  size?: ButtonSize;
};

function elevationShadowStyle(elevation: number) {
  return {
    elevation,
    shadowColor: 'black',
    shadowOffset: { width: 0, height: 0.5 * elevation },
    shadowOpacity: 0.3,
    shadowRadius: 0.8 * elevation,
  };
}

function withSizes(size: ButtonSize | undefined = 'medium', theme: any) {
  const { buttonSize }: { buttonSize: IButtonSize } = theme;
  if (!buttonSize) return {};

  return buttonSize[size];
}

const Button: React.FunctionComponent<Omit<ButtonProps, 'height' | 'width'>> = (
  props: Omit<ButtonProps, 'height' | 'width'>
) => {
  const {
    theme,
    icon,
    colorType,
    mode,
    elevation,
    disabled,
    size: propsSize,
  } = props;

  const themeColors = {
    primary: theme.colors.primary,
    secondary: theme.colors.secondary,
    disabled: theme.colors.disabled,
    text: theme.colors.text,
    default: '#E0E0E0',
  };

  const margin = {
    default: 0,
    reduceFromIcon: -4,
  };

  const color = React.useMemo(() => {
    let colorReturn = themeColors.primary;

    if (colorType) {
      colorReturn = themeColors[colorType];
    }

    if (mode === 'outlined' || (mode === 'text' && colorType === 'default')) {
      colorReturn = themeColors.text;
    }

    if (disabled) {
      colorReturn = themeColors.disabled;
    }

    return colorReturn;
  }, [colorType, mode, disabled]);

  const borderColor = React.useMemo(() => {
    let borderColorReturn = theme.colors.primary;

    if (colorType) {
      borderColorReturn = themeColors[colorType];
    }

    if (disabled) {
      borderColorReturn = themeColors.disabled;
    }

    return borderColorReturn;
  }, [colorType, disabled]);

  const style = React.useMemo(() => {
    return elevation && !disabled ? elevationShadowStyle(elevation) : {};
  }, [elevation, disabled]);

  const labelStyleMargin = React.useMemo(
    () => ({
      marginTop: 0,
      marginRight: 0,
      marginBottom: 0,
      marginLeft: icon ? margin.reduceFromIcon : margin.default,
    }),
    [icon]
  );

  const contentStyleMargin = React.useMemo(
    () => ({
      marginLeft:
        icon && propsSize === 'small' ? margin.reduceFromIcon : margin.default,
    }),
    [icon, propsSize]
  );

  const size = React.useMemo(() => withSizes(propsSize, theme), [
    propsSize,
    theme,
  ]);

  return (
    <PaperButton
      uppercase={true}
      {...props}
      color={color}
      style={{ ...style, borderColor }}
      contentStyle={{ ...contentStyleMargin }}
      labelStyle={{
        ...size,
        ...labelStyleMargin,
        fontWeight: theme.fonts.regular.fontWeight,
      }}
    />
  );
};

export default withTheme(Button);
