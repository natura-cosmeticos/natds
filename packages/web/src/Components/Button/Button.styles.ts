import makeStyles from '@material-ui/core/styles/makeStyles'
import createStyles from '@material-ui/core/styles/createStyles'
import { IThemeWeb } from '../../Themes'

type ButtonRootStyles = {
  paddingLeft?: number,
  paddingRight?: number,
}

type ButtonLabelStyles = {
  lineHeight?: string,
  height?: number,
  marginBottom?: number,
  marginTop?: number
};

type ButtonStyles = {
  root: ButtonRootStyles,
  label: ButtonLabelStyles
}

export const getRootStyles: (theme: IThemeWeb) => (ButtonRootStyles) = (theme) => {
  if (typeof theme === 'undefined' || typeof theme.sizes === 'undefined' || typeof theme.sizes.small === 'undefined') {
    return {} as ButtonRootStyles
  }

  return {
    paddingLeft: theme?.sizes?.small,
    paddingRight: theme?.sizes?.small
  } as ButtonRootStyles
}

export const getLabelStyles: (theme: IThemeWeb) => (ButtonLabelStyles) = (theme) => {
  if (typeof theme === 'undefined' || typeof theme.sizes === 'undefined') {
    return {} as ButtonLabelStyles
  }

  return {
    lineHeight: `${theme.sizes.standard}px`,
    height: theme.sizes?.standard,
    marginBottom: theme.sizes.tiny,
    marginTop: theme.sizes.tiny
  } as ButtonLabelStyles
}

export const styles: (theme: IThemeWeb) => ButtonStyles = (theme) => ({
  root: getRootStyles(theme),
  label: getLabelStyles(theme)
} as ButtonStyles)

export const useStyles = makeStyles(createStyles(styles), { name: 'NatDSButton' })

export default useStyles
