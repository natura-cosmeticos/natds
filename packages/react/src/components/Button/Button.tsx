import React from 'react'
import { ButtonProps } from './Button.props'
import { Ripple } from '../Ripple'
import styles from './Button.styles'

const Button = ({
  disabled = false,
  fullWidth = false,
  label,
  onClick,
  size = 'semiX',
  testID,
  variant = 'contained'
}: ButtonProps): JSX.Element => {
  const classes = styles({
    size, fullWidth, variant, disabled
  })

  return (
    <Ripple disabled={disabled} fullWidth={fullWidth}>
      <button
        className={classes.button}
        data-testid={testID}
        disabled={disabled}
        onClick={onClick}
        type="button"
      >
        <span className={classes.label}>{label}</span>
      </button>
    </Ripple>
  )
}

export default Button
