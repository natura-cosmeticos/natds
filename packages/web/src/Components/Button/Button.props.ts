import { ButtonProps } from '@material-ui/core/Button'
import { PropTypes } from '@material-ui/core'
import * as React from 'react'

export type ButtonColor = PropTypes.Color;
export type ButtonSize = 'small' | 'medium' | 'large';
export type ButtonVariant = 'text' | 'outlined' | 'contained';

export interface IButtonProps extends Omit<ButtonProps, 'translate'> {

  /**
   * The content of the button
   *
   * @optional
   * @type node
   */
  children?: ButtonProps['children']

  /**
   * Override or extend the styles applied to the component.
   *
   * See [Material Button CSS API](https://material-ui.com/api/button/#css) for more details.
   *
   * @optional
   * @see https://material-ui.com/api/button/#css
   * @type object
   */
  classes?: ButtonProps['classes']

  /**
   * The color of the `Button` component. It supports those theme colors that make sense for this component.
   *
   * @default "default"
   * @optional
   * @type "default" | "inherit" | "primary" | "secondary"
   */
  color?: ButtonProps['color']

  /**
   * The component used for the root node. Either a `string` to use a DOM element or a component.
   *
   * @default "button"
   * @optional
   * @type element
   */
  component?: React.ReactElement

  /**
   *
   * If `true`, the button will be disabled.
   *
   * @default false
   * @type bool
   */
  disabled?: ButtonProps['disabled']

  textTransform?:'uppercase' | 'lowercase' | 'capitalize'

  /**
   * If `true`, no elevation is used.
   *
   * @default false
   * @optional
   * @type bool
   */
  disableElevation?: ButtonProps['disableElevation']

  /**
   * If `true`, the keyboard focus ripple will be disabled. `disableRipple` must also be `true`.
   *
   * @default false
   * @optional
   * @type bool
   */
  disableFocusRipple?: ButtonProps['disableFocusRipple']

  /**
   * If `true`, the ripple effect will be disabled.
   *
   * ⚠️ Without a ripple there is no styling for :focus-visible by default. Be sure
   * to highlight the element by applying separate styles with the `focusVisibleClassName`.
   *
   * @optional
   * @type bool
   */
  disableRipple?: ButtonProps['disableRipple']

  /**
   * Element placed after the children.
   *
   * @optional
   * @type node
   */
  endIcon?: ButtonProps['endIcon']

  /**
   * If `true`, the `Button` will take up the full width of its container.
   *
   * @default false
   * @optional
   * @type bool
   */
  fullWidth?: ButtonProps['fullWidth'];

  /**
   * The URL to link to when the button is clicked.
   * If defined, an `a` element will be used as the root node.
   *
   * @optional
   * @type string
   */
  href?: ButtonProps['href']

  /**
   * The size of the button.
   * `small` is equivalent to the dense button styling.
   *
   * @todo check if "small" size is still on UI Specs
   * @type "small" | "medium" | "large"
   */
  size?: ButtonProps['size']

  /**
   * Element placed before the children.
   *
   * @optional
   * @type node
   */
  startIcon?: ButtonProps['startIcon']

  /**
   * The variant to use.
   *
   * @optional
   * @type "contained" | "outlined" | "text"
   */
  variant?: ButtonProps['variant']
}
