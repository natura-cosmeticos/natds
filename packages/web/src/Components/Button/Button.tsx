import * as React from 'react'

import MaterialButton from '@material-ui/core/Button'
import { IButtonProps } from './Button.props'
import useStyles from './Button.styles'

export { IButtonProps } from './Button.props'

/**
 * ```
 * import { Button } from '@naturacosmeticos/natds-web';
 * ```
 *
 * For more advanced usages, check [Material UI Button docs](https://material-ui.com/components/buttons/)
 */
export const Button = React.forwardRef<HTMLButtonElement, IButtonProps>(
  (props: IButtonProps, ref) => {
    const { classes, disableElevation, ...otherProps } = props

    const mergedClasses = {
      ...classes,
      ...useStyles()
    }

    return (
      <MaterialButton
        classes={mergedClasses}
        disableElevation={disableElevation}
        ref={ref}
        {...otherProps}
      />
    )
  }
)

Button.displayName = 'Button'
export default Button
