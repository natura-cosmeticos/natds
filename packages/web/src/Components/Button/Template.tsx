import { Story } from '@storybook/react'
import * as React from 'react'
import { IButtonProps } from './Button.props'
import { Button } from './Button'

export const Template : Story<IButtonProps> = (args) => <Button {...args} />

export const TemplateWithColors : Story<IButtonProps> = (args) => (
  <>
    <Button {...args} color="primary">Primary color</Button>
&nbsp;
    <Button {...args} color="secondary">Secondary color</Button>
&nbsp;
    <Button {...args} color="default">Default color</Button>
&nbsp;
  </>
)

export const TemplateWithSizes : Story<IButtonProps> = (args) => (
  <>
    <Button {...args} size="small">Small size</Button>
&nbsp;
    <Button {...args} size="medium">Medium size</Button>
&nbsp;
    <Button {...args} size="large">Large size</Button>
&nbsp;
  </>
)

export default Template
