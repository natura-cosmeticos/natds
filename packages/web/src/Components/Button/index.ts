import Button from './Button'

export { IButtonProps } from './Button.props'
export { Button }
export default Button
