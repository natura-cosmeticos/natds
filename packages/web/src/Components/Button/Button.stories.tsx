import * as React from 'react'
import { Story } from '@storybook/react'
import { Button, IButtonProps } from './Button'
import { Icon } from '../Icon'
import { colors } from './__fixtures__/colors'
import { sizes } from './__fixtures__/sizes'
import { variants } from './__fixtures__/variants'
import { Template, TemplateWithColors, TemplateWithSizes } from './Template'

export default { component: Button, title: 'Components/Button' }

export const Playground : Story<IButtonProps> = Template.bind({})
Playground.args = {
  children: 'Button component text',
  color: colors.primary,
  endIcon: false,
  size: sizes.medium,
  startIcon: false,
  variant: variants.contained
}

export const ContainedVariant : Story<IButtonProps> = Template.bind({})
ContainedVariant.args = {
  ...Playground.args,
  variant: 'contained'
}

export const OutlinedVariant : Story<IButtonProps> = Template.bind({})
OutlinedVariant.args = {
  ...Playground.args,
  variant: 'outlined'
}

export const TextVariant : Story<IButtonProps> = Template.bind({})
TextVariant.args = {
  ...Playground.args,
  variant: 'text'
}

export const Disabled : Story<IButtonProps> = Template.bind({})
Disabled.args = {
  ...Playground.args,
  disabled: true
}

export const StartIcon : Story<IButtonProps> = Template.bind({})
StartIcon.args = {
  ...Playground.args,
  startIcon: <Icon name="outlined-action-favorite" />
}

export const EndIcon : Story<IButtonProps> = Template.bind({})
EndIcon.args = {
  ...Playground.args,
  endIcon: <Icon name="outlined-action-favorite" />
}

export const Colors : Story<IButtonProps> = TemplateWithColors.bind({})
Colors.args = Playground.args

export const InheritColor : Story<IButtonProps> = Template.bind({})
InheritColor.args = {
  ...Playground.args,
  color: colors.inherit
}
InheritColor.parameters = {
  docs: {
    description: {
      story: "The `inherit` color is useful when you need to apply a button to an `AppBar`'s `Toolbar`."
    }
  }
}

export const ForSnackbar : Story<IButtonProps> = Template.bind({})
ForSnackbar.args = {
  ...Playground.args,
  color: colors.inherit,
  variant: variants.text
}

export const ForDialogActions : Story<IButtonProps> = Template.bind({})
ForDialogActions.args = {
  children: 'Action text',
  color: colors.primary,
  variant: variants.text
}

export const Sizes : Story<IButtonProps> = TemplateWithSizes.bind({})
Sizes.args = Playground.args

/**
 * @todo Check if small size is still part of UI specs
 */
export const SmallSize : Story<IButtonProps> = Template.bind({})
SmallSize.args = {
  ...Playground.args,
  size: sizes.small
}

export const MediumSize : Story<IButtonProps> = Template.bind({})
MediumSize.args = {
  ...Playground.args,
  size: sizes.medium
}

export const LargeSize : Story<IButtonProps> = Template.bind({})
LargeSize.args = {
  ...Playground.args,
  size: sizes.large
}

export const FullWidth : Story<IButtonProps> = TemplateWithSizes.bind({})
FullWidth.args = {
  ...Playground.args,
  fullWidth: true
}

export const ForActions : Story<IButtonProps> = TemplateWithSizes.bind({})
ForActions.args = {
  children: 'Button label',
  size: sizes.small
}
