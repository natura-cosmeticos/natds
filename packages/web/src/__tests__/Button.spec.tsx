import React from 'react';
import { shallow } from 'enzyme';
import { expect } from 'chai';
import { assert, spy } from 'sinon';
import renderer from 'react-test-renderer';

import Button from '../Components/Button';

const label = 'A Button Component';

describe('Button component', () => {
  describe('rendering sizes', () => {
    test('should match snapshot - Size Small', () => {
      const component = renderer.create(<Button size="small">{label}</Button>).toJSON();
      expect(component).matchSnapshot('Small Button snapshot');
    });

    test('should match snapshot - Size Medium', () => {
      const component = renderer.create(<Button size="medium">{label}</Button>).toJSON();
      expect(component).matchSnapshot('Medium Button snapshot');
    });

    test('should match snapshot - Size Large', () => {
      const component = renderer.create(<Button size="large">{label}</Button>).toJSON();
      expect(component).matchSnapshot('Medium Button snapshot');
    });
  });

  describe('rendering variants', () => {
    test('should match snapshot - Low emphasis button (variant text)', () => {
      const component = renderer.create(<Button variant="text">{label}</Button>).toJSON();

      expect(component).to.matchSnapshot('Low emphasis Button snapshot');
    });

    test('should match snapshot - Medium emphasis button (variant outlined)', () => {
      const component = renderer.create(<Button variant="outlined">{label}</Button>).toJSON();

      expect(component).to.matchSnapshot('Medium emphasis Button snapshot');
    });

    test('should match snapshot - High emphasis button (variant contained)', () => {
      const component = renderer.create(<Button variant="contained">{label}</Button>).toJSON();

      expect(component).to.matchSnapshot('High emphasis Button snapshot');
    });
  });

  describe('interaction', () => {
    test('should call onClick', () => {
      const mockOnClick = spy();
      const component = shallow(<Button onClick={mockOnClick} />);

      component.simulate('click');

      assert.calledOnce(mockOnClick);
    });
  });
});
