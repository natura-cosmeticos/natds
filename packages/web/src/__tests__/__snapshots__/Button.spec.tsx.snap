// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component rendering sizes should match snapshot - Size Large: Large Button snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeLarge MuiButton-sizeLarge"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    A Button Component
  </span>
</button>
`;

exports[`Button component rendering sizes should match snapshot - Size Medium: Medium Button snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    A Button Component
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button component rendering sizes should match snapshot - Size Small: Small Button snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSizeSmall MuiButton-sizeSmall"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    A Button Component
  </span>
</button>
`;

exports[`Button component rendering variants should match snapshot - High emphasis button (variant contained): High emphasis Button snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    A Button Component
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button component rendering variants should match snapshot - Low emphasis button (variant text): Low emphasis Button snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    A Button Component
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;

exports[`Button component rendering variants should match snapshot - Medium emphasis button (variant outlined): Medium emphasis Button snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    A Button Component
  </span>
</button>
`;
